### PR TITLE
azure-sdk-for-cpp/1.11.3:new recipe

### PR DESCRIPTION
**Pull request by trns1997**
Add azure-sdk-for-cpp https://github.com/Azure/azure-sdk-for-cpp as the old azure-storage-cpp (homepage = "https://github.com/Azure/azure-storage-cpp) is deprecated (conan deprecated package: https://github.com/conan-io/conan-center-index/pull/18842). Also couple of users requested to add this package to conan: https://github.com/Azure/azure-sdk-for-cpp/issues/5485

For the moment I mainly focused on building the entire sdk and only on Linux. Windows and apple not supported yet, contributions welcomed!

---

- [x] I've read the [contributing guidelines](https://github.com/conan-io/conan-center-index/blob/master/CONTRIBUTING.md).
- [x] I've used a [recent](https://github.com/conan-io/conan/releases/latest) Conan client version close to the [currently deployed](https://github.com/conan-io/conan-center-index/blob/master/.c3i/config_v1.yml#L6).
- [x] I've tried at least one configuration locally with the [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

File: recipes/azure-sdk-for-cpp/all/conanfile.py [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+    def source(self):
+        get(self, **self.conan_data["sources"][self.version], strip_root=True)
+
+    def configure(self):
+        if self.options.get_safe("shared"):
+            self.options.rm_safe("fPIC")
+
+    def requirements(self):
+        self.requires("libcurl/8.6.0")
+        self.requires("libxml2/2.12.6")
```
Comment (RubenRBS) [inline recipes/azure-sdk-for-cpp/all/conanfile.py:49]:
```suggestion
        self.requires("libcurl/[>=7.78 <9]")
        self.requires("libxml2/[>=2.12.5 <3]")
```

File: recipes/azure-sdk-for-cpp/all/conanfile.py [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+        if self.settings.compiler.get_safe("cppstd"):
+            check_min_cppstd(self, 14)
+
+        if self.settings.os != "Linux":
+            raise ConanInvalidConfiguration(
+                f"{self.ref} is not supported on {self.settings.os}.")
+
+        if self.settings.compiler != "gcc":
+            raise ConanInvalidConfiguration(
+                f"{self.ref} is not supported on {self.settings.compiler}.")
```
Comment (RubenRBS) [inline recipes/azure-sdk-for-cpp/all/conanfile.py:64]: By the wording on your PR comment
> and only on Linux.

it seems these limitations are a recipe one, and not something coming from upstream.

If that's the case, I'd suggest changing the wording here to add something like "not supported yet, contributions welcomed!" or something like that :)

File: recipes/azure-sdk-for-cpp/all/test_package/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,9 @@
+cmake_minimum_required(VERSION 3.8)
+project(test_package LANGUAGES CXX)
+
+set(CMAKE_CXX_STANDARD 17)
```
Comment (RubenRBS) [inline recipes/azure-sdk-for-cpp/all/test_package/CMakeLists.txt:4]: Usually we prefer the min cxx version to match the one declared in the recipe itself. Any specific reason this is not in this case?

**Review by RubenRBS — COMMENTED** [review]
Hi! Thanks a lot for your contribution, we appreciate it

I took a quick look, will be able to have a better review next monday, but for now I've found a few minor things worth commenting on :)

File: recipes/azure-sdk-for-cpp/all/conanfile.py (hunk 2 above)
Comment (trns1997) [reply recipes/azure-sdk-for-cpp/all/conanfile.py:64]: Clearly that is what i meant. I have updated the comment saying that windows and apple support is not yet donee and contributions welcomes :).

File: recipes/azure-sdk-for-cpp/all/test_package/CMakeLists.txt (hunk 3 above)
Comment (trns1997) [reply recipes/azure-sdk-for-cpp/all/test_package/CMakeLists.txt:4]: ah not at all nice catch i will, i based it off a recipe i made before and copied it blindly.

**Comment by trns1997** [comment]
@RubenRBS I would like to get your inputs on this PR. I would like to install this package via Conan

**Comment by RubenRBS** [comment]
Hi @trns1997 , sorry for the delay, most of the team was attending a conference last week which delayed reviews a bit, will get back to you tomorrow :)

**Comment by trns1997** [comment]
@RubenRBS I created a cmake wrapper allowing us to build azure as modules. I would like your input before i go on to clean it up, add all modules of the sdk and add tests for each module.

File: recipes/azure-sdk-for-cpp/all/conanfile.py [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+            self.cpp_info.components[sdk].names["cmake_find_package"] = sdk
+            self.cpp_info.components[sdk].names["cmake_find_package_multi"] = sdk
+
+        # TODO: to remove in conan v2 once cmake_find_package_* generators removed
+        self.cpp_info.filenames["cmake_find_package"] = "AzureSDK"
+        self.cpp_info.filenames["cmake_find_package_multi"] = "AzureSDK"
+        self.cpp_info.names["cmake_find_package"] = "Azure"
+        self.cpp_info.names["cmake_find_package_multi"] = "Azure"
+        self.cpp_info.components["azure-core"].names["cmake_find_package"] = "azure-core"
+        self.cpp_info.components["azure-core"].names["cmake_find_package_multi"] = "azure-core"
```
Comment (memsharded) [inline recipes/azure-sdk-for-cpp/all/conanfile.py:122]: These are no longer necessary for new recipes, can be dropped

File: recipes/azure-sdk-for-cpp/all/conanfile.py [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+        tc.cache_variables["DISABLE_AZURE_CORE_OPENTELEMETRY"] = "ON"
+        tc.cache_variables["BUILD_TRANSPORT_CURL"] = "ON"
+        tc.generate()
+
+        deps = CMakeDeps(self)
+        deps.generate()
+
+    def build(self):
+        cmake = CMake(self)
+        cmake.configure(build_script_folder=self.export_sources_folder)
```
Comment (memsharded) [inline recipes/azure-sdk-for-cpp/all/conanfile.py:87]: This sounds problematic. The build shouldn't use the ``export_sources_folder``. Those files are already copied into the ``source_folder`` and depending on the ``no_copy_sources``, to the ``build_folder`` too.

File: recipes/azure-sdk-for-cpp/all/conanfile.py (hunk 2 above)
Comment (memsharded) [reply recipes/azure-sdk-for-cpp/all/conanfile.py:60]: I am still not convinced. If I read ``"azure-sdk-for-cpp is not supported on Windows"`` I still think that the library doesn't support Windows. I think it should be more explicit: "The azure-sdk-for-cpp Conan recipe in ConanCenter still does not support Windows, contributions to the recipe welcome"

File: recipes/azure-sdk-for-cpp/all/CMakeLists.txt [hunk 6]
```diff
@@ -0,0 +1,15 @@
+cmake_minimum_required(VERSION 3.15)
+project(cmake_wrapper)
+
+foreach(sdk ${BUILD_LIST})
+    if(${sdk} STREQUAL azure-core)
+        # Always build Core
+        add_subdirectory("src/sdk/core")
+    elseif(${sdk} STREQUAL azure-storage-common)
```
Comment (memsharded) [inline recipes/azure-sdk-for-cpp/all/CMakeLists.txt:8]: A few lines explaining why this CMakeLists.txt needs to be injected in the build would be great.

**Review by memsharded — COMMENTED** [review]
The recipe reads nice and simple overall, just a few minor comments.

**Comment by trns1997** [comment]
@RubenRBS @memsharded ready for review :)

**Comment by trns1997** [comment]
@RubenRBS @memsharded, can you guys reviews when you have the time :)

**Review by memsharded — APPROVED** [review]

**Review by AbrilRBS — APPROVED** [review]

**Pull request merged**